### PR TITLE
feat: Add a validation  init container

### PR DESCRIPTION
**Pull request by helayoty**
- Create Dockerfile and pkg folder for init container
- Add support for init container in helm chart, deployments for e2e tests
- Add support for init container in pipelines 

Signed-off-by: Heba Elayoty <hebaelayoty@gmail.com>

### CODE REVIEW

File: cmd/init-container/main.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+			log.G(ctx).Fatal(err)
+		}
+	}
+	p := provider.ACIProvider{
+		ProviderNetwork: network.ProviderNetwork{},
+	}
+	// Check or set up a network for VK
+	err = p.ProviderNetwork.SetVNETConfig(ctx, &azConfig)
+	if err != nil {
+
```
Comment (Fei-Guo) [inline cmd/init-container/main.go:44]: What do we do for the error msg?  I think this is the key for the init container to provide sufficient log information for how to fixing errors and providing remediations.

Comment (helayoty) [reply cmd/init-container/main.go]: Added more tracing and logs.

File: cmd/init-container/main.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	config, err := clientcmd.BuildConfigFromFlags("", "")
+	if err != nil {
+		log.G(ctx).Fatal("an error has occurred while creating client ", err)
+	}
+
+	kubeClient := kubernetes.NewForConfigOrDie(config)
+	eventBroadcast := util.NewRecorder(ctx, kubeClient)
+	defer eventBroadcast.Shutdown()
+
+	recorder := eventBroadcast.NewRecorder(scheme.Scheme, v1.EventSource{Component: "k8s.io.api.core.v1.Pod"})
```
Comment (Fei-Guo) [inline cmd/init-container/main.go:54]: I think the component should be the agent name that fires the event, so it should be "virtual kubelet" or something similar.

File: cmd/init-container/main.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+		})
+
+	if err != nil {
+		recorder.Eventf(&v1.ObjectReference{
+			Kind:      "Pod",
+			Name:      podName,
+			Namespace: podNamespace,
+		}, v1.EventTypeWarning, "InitFailed", "VNet config setup failed")
+		log.G(ctx).Fatal("cannot setup the VNet configuration ", err)
+	}
```
Comment (Fei-Guo) [inline cmd/init-container/main.go:102]: You should add a `return` at line 102.

File: pkg/util/events.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+	"k8s.io/client-go/kubernetes"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
+)
+
+func NewRecorder(ctx context.Context, kubeClient *kubernetes.Clientset) record.EventBroadcaster {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartStructuredLogging(3)
+	if eventBroadcaster != nil && kubeClient != nil {
```
Comment (Fei-Guo) [inline pkg/util/events.go:18]: Can eventBroadcaster be nil in this case?

File: pkg/network/aci_network.go [hunk 5]
```diff
@@ -42,36 +43,49 @@ type ProviderNetwork struct {
 }
 
 func (pn *ProviderNetwork) SetVNETConfig(ctx context.Context, azConfig *auth.Config) error {
+	ctx, span := trace.StartSpan(ctx, "network.SetVNETConfig")
+	defer span.End()
+
+	if azConfig.AKSCredential != nil {
```
Comment (Fei-Guo) [inline pkg/network/aci_network.go:49]: Is this a new behavior?

Comment (helayoty) [reply pkg/network/aci_network.go:49]: No, this is only to add a trace for the 'SetVNETConfig' method.

File: cmd/init-container/main.go (hunk 3 above)
Comment (helayoty) [reply cmd/init-container/main.go:102]: The Fatal will execute os.Exit(1) as implemented in the [logrus](https://github.com/sirupsen/logrus/blob/master/entry.go#L383) library.

File: pkg/util/events.go (hunk 4 above)
Comment (helayoty) [reply pkg/util/events.go]: No need. Removed.

File: cmd/init-container/main.go (hunk 2 above)
Comment (helayoty) [reply cmd/init-container/main.go]: Updated.

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**